File: packages/react-native-nitro-test/ios/HybridPlatformObject.swift
//
//  HybridPlatformObject.swift
//  react-native-nitro-test
//
//  Created by Marc Rousavy on 27.10.25.
//

import UIKit

struct ReactModuleAccessError: Error, CustomStringConvertible {
  let description: String
}

class HybridPlatformObject: HybridPlatformObjectSpec {
  func getOSVersion() -> String {
    return UIDevice.current.systemVersion
  }

  // Get the RN "BlobModule" from inside a Nitro HybridObject and use it.
  // Android: `NitroModules.applicationContext.getNativeModule(BlobModule::class.java)`.
  // iOS: there is no `NitroModules.applicationContext`; the only entry point is the
  // private `RCTBridge.currentBridge` (called dynamically below), which is nil under
  // the New Architecture (bridgeless) — so there is nothing to call.
  func getReactContextInfo() throws -> String {
    let bridge = (NSClassFromString("RCTBridge") as AnyObject?)?
      .perform(NSSelectorFromString("currentBridge"))?
      .takeUnretainedValue()
    guard let bridge else {
      throw ReactModuleAccessError(
        description: "RCTBridge.currentBridge == nil (bridgeless): a Nitro HybridObject cannot reach BlobModule or any RN native module on iOS")
    }
    let blobModule = (bridge as AnyObject)
      .perform(NSSelectorFromString("moduleForName:"), with: "BlobModule")?
      .takeUnretainedValue()
    return "iOS reached BlobModule = \(String(describing: blobModule))"
  }
}

File: packages/react-native-nitro-test/android/src/main/java/com/margelo/nitro/test/HybridPlatformObject.kt
package com.margelo.nitro.test

import android.os.Build
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.modules.blob.BlobModule
import com.margelo.nitro.NitroModules

@Keep
@DoNotStrip
class HybridPlatformObject : HybridPlatformObjectSpec() {
  override fun getOSVersion(): String {
    return Build.VERSION.RELEASE
  }

  // Android CAN reach an RN native module from a HybridObject via
  // NitroModules.applicationContext — iOS has no equivalent.
  override fun getReactContextInfo(): String {
    val ctx = NitroModules.applicationContext
      ?: throw IllegalStateException("NitroModules.applicationContext == null")
    val blobModule = ctx.getNativeModule(BlobModule::class.java)
      ?: throw IllegalStateException("BlobModule not found")
    val blobId = blobModule.store(byteArrayOf(1, 2, 3))
    return "Android reached BlobModule, store([1,2,3]) -> $blobId"
  }
}
